### ISSUE TITLE
github.com/gobwas/glob appears to be unmaintained

### ISSUE THREAD

**Issue by charlesdaniels** [open-policy-agent/opa#4780]
I have been looking into some glob matching stuff lately (see #4777). I noticed we are using `github.com/gobwas/glob` to implement our glob matching, and indeed this seems to be the only open source Go library that suits our needs. However, the project in question appears to be mostly abandoned, with no commits since 2018, and no activity at all in over a year. We should consider one of the following options:

* Adopt/fork `gobwas/glob` and maintain it ourselves going forward
* Write our own glob matching library and use/maintain that going forward

I think the latter option may be the way to go. I don't think this would be too hard, as `github.com/gobwas/glob` is only around 5k SLoC, and also does it's own compilation/execution as far as I can tell. A lot of modern languages these days implement globbing by converting the glob patterns to regular expressions and then executing those ([source](https://research.swtch.com/glob)), and I don't think that would be too challenging in our case. This would also allow us to eliminate a non standard library dependency.

I don't think this is a top priority, as there don't appear to be known problems with `github.com/gobwas/glob` today. Based on my investigation, I don't believe there is a ton of performance laying around on the floor ready to be picked up.

**Comment by srenatus** [comment]
> However, the project in question appears to be mostly abandoned, with no commits since 2018, and no activity at all in over a year. We should consider one of the following options:
> 
> * Adopt/fork gobwas/glob and maintain it ourselves going forward
> * Write our own glob matching library and use/maintain that going forward

As for all third-party code we're using -- copied into our sources or imported the standard golang modules way -- "adopt" is the implicit first choice: *When importing something, we make it our problem.* For an OPA user it's irrelevant what code backs a built-in, and they'll expect it to work in the documented way. (And that's all we expect from them, not some sort of negligence on the users' part.)

For the library in question, though, I wonder what problem we're trying to solve. If the library works well for us today, and there are no known bugs, or other sorts of issued that need addressing, I don't think the time of last commit matters at all. Looking at the open (neglected) PRs of gobwas/glob, there's nothing standing out as critical, I think. Looking at our open issues, I don't see any, either.

Bottom line, I don't think there's anything to do here.

As for the performance, if you'd like to look into approaching the backing code of our glob built-ins in a more modern way, feel free to do so. I'm sure nobody clings to this particular third party dependency. All I'm saying is that I think you're correct, there's nothing urgent to do here.

**Comment by anderseknert** [comment]
I agree with Stephan here. On the other side of this, there are the libraries that _are_ maintained but we "forked", or rather included in OPA by copying code — like the jwx one — where we've had to make modifications over the years to meet requirements in the OPA project, leading to us not automatically getting updates from the upstream library, and doing that manually would entail quite some effort. Unless there are known issues that won't be addressed, having stable/frozen dependencies isn't all that bad :)

**Comment by srenatus** [comment]
@charlesdaniels don't get me wrong, I really appreciate you looking into stuff like that and sharing the results. Thank you!
